File: public-api/examples/python/batch-search.py
"""
Batch Search Example

This example demonstrates how to search for multiple queries in a single request,
which is more efficient than making multiple sequential calls.
"""

import asyncio
import os
import httpx


async def batch_search(queries: list[str], api_key: str) -> dict:
    """
    Perform batch search for multiple queries.
    
    Args:
        queries: List of search queries
        api_key: Your Ad-Tokens API key
        
    Returns:
        Dictionary containing batch search results
    """
    async with httpx.AsyncClient() as client:
        try:
            # Prepare batch request
            batch_request = {
                "queries": [
                    {"query": query, "limit": 3}
                    for query in queries
                ]
            }
            
            response = await client.post(
                "https://api.ad-tokens.com/search/batch",
                headers={
                    "x-api-key": api_key,
                    "Content-Type": "application/json",
                },
                json=batch_request,
                timeout=30.0,
            )
            response.raise_for_status()
            return response.json()
            
        except httpx.HTTPStatusError as e:
            print(f"❌ HTTP error: {e.response.status_code}")
            print(f"   Error: {e.response.text}")
            raise
        except httpx.RequestError as e:
            print(f"❌ Request error: {e}")
            raise


async def main():
    api_key = os.getenv("AD_TOKENS_API_KEY", "your-api-key-here")
    
    if api_key == "your-api-key-here":
        print("⚠️  Please set AD_TOKENS_API_KEY environment variable")
        return
    
    # Multiple queries to search
    queries = [
        "wireless headphones",
        "mechanical keyboard",
        "gaming mouse",
    ]
    
    print(f"🔍 Batch searching for {len(queries)} queries...\n")
    
    try:
        result = await batch_search(queries, api_key)
        
        # Display results for each query
        for i, search_result in enumerate(result["results"], 1):
            query = queries[i - 1]
            print(f"📦 Query {i}: {query}")
            print(f"   Found {len(search_result['results'])} products\n")
            
            for j, product in enumerate(search_result["results"], 1):
                print(f"   {j}. {product['title']}")
                print(f"      Price: {product['price']}")
                print(f"      Relevance: {product['relevance_score']:.2%}")
            
            print()
        
        # Display batch metadata
        metadata = result.get("metadata", {})
        if "total_queries" in metadata:
            print(f"✅ Processed {metadata['total_queries']} queries")
        if "total_time_ms" in metadata:
            print(f"⏱️  Total time: {metadata['total_time_ms']:.2f}ms")
            
    except Exception as e:
        print(f"❌ Error: {e}")


if __name__ == "__main__":
    asyncio.run(main())

File: public-api/examples/javascript/batch-search.js
/**
 * Batch Search Example
 * 
 * This example demonstrates how to search for multiple queries in a single request,
 * which is more efficient than making multiple sequential calls.
 */

/**
 * Perform batch search for multiple queries.
 * 
 * @param {string[]} queries - Array of search queries
 * @param {string} apiKey - Your Ad-Tokens API key
 * @returns {Promise<Object>} Batch search results
 */
async function batchSearch(queries, apiKey) {
  try {
    // Prepare batch request
    const batchRequest = {
      queries: queries.map(query => ({
        query: query,
        limit: 3,
      })),
    };

    const response = await fetch('https://api.ad-tokens.com/search/batch', {
      method: 'POST',
      headers: {
        'x-api-key': apiKey,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(batchRequest),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(`HTTP error! status: ${response.status}, message: ${error.message}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Batch search error:', error);
    throw error;
  }
}

/**
 * Main function to demonstrate batch search.
 */
async function main() {
  const apiKey = process.env.AD_TOKENS_API_KEY || 'your-api-key-here';

  if (apiKey === 'your-api-key-here') {
    console.error('⚠️  Please set AD_TOKENS_API_KEY environment variable');
    return;
  }

  // Multiple queries to search
  const queries = [
    'wireless headphones',
    'mechanical keyboard',
    'gaming mouse',
  ];

  console.log(`🔍 Batch searching for ${queries.length} queries...\n`);

  try {
    const result = await batchSearch(queries, apiKey);

    // Display results for each query
    result.results.forEach((searchResult, index) => {
      const query = queries[index];
      console.log(`📦 Query ${index + 1}: ${query}`);
      console.log(`   Found ${searchResult.results.length} products\n`);

      searchResult.results.forEach((product, productIndex) => {
        console.log(`   ${productIndex + 1}. ${product.title}`);
        console.log(`      Price: ${product.price}`);
        console.log(`      Relevance: ${(product.relevance_score * 100).toFixed(2)}%`);
      });

      console.log();
    });

    // Display batch metadata
    if (result.metadata) {
      if (result.metadata.total_queries) {
        console.log(`✅ Processed ${result.metadata.total_queries} queries`);
      }
      if (result.metadata.total_time_ms) {
        console.log(`⏱️  Total time: ${result.metadata.total_time_ms.toFixed(2)}ms`);
      }
    }
  } catch (error) {
    console.error(`❌ Error: ${error.message}`);
  }
}

// Run if this is the main module
if (typeof require !== 'undefined' && require.main === module) {
  main();
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { batchSearch };
}

